[SPARK-5063] More helpful error messages for several invalid operations

JoshRosen opened the pull request:
This patch adds more helpful error messages for invalid programs that define nested RDDs, broadcast RDDs, perform actions inside of transformations (e.g. calling `count()` from inside of `map()`), and call certain methods on stopped SparkContexts.  Currently, these invalid programs lead to confusing NullPointerExceptions at runtime and have been a major source of questions on the mailing list and StackOverflow.

In a few cases, I chose to log warnings instead of throwing exceptions in order to avoid any chance that this patch breaks programs that worked "by accident" in earlier Spark releases (e.g. programs that define nested RDDs but never run any jobs with them).

In SparkContext, the new `assertNotStopped()` method is used to check whether methods are being invoked on a stopped SparkContext.  In some cases, user programs will not crash in spite of calling methods on stopped SparkContexts, so I've only added `assertNotStopped()` calls to methods that always throw exceptions when called on stopped contexts (e.g. by dereferencing a null `dagScheduler` pointer).

JoshRosen committed:
[SPARK-5063] Useful error messages for nested RDDs and actions inside of transformations

SparkQA commented:
[Test build #25005 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25005/consoleFull) for   PR 3884 at commit [`15b2e6b`](https://github.com/apache/spark/commit/15b2e6b38d9587790357182abe7918853688722e).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25005 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25005/consoleFull) for   PR 3884 at commit [`15b2e6b`](https://github.com/apache/spark/commit/15b2e6b38d9587790357182abe7918853688722e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25005/
Test PASSed.

sryza commented:
Will this work for broadcast variables as well?  One thing I often see is users trying to directly broadcast an RDD without collecting it.

JoshRosen committed:
Add error message when directly broadcasting RDD.

SparkQA commented:
[Test build #25036 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25036/consoleFull) for   PR 3884 at commit [`57cc8a1`](https://github.com/apache/spark/commit/57cc8a11266770e145c7ca810bec3b95aeefabb3).
 * This patch merges cleanly.

JoshRosen commented:
@sryza Good idea; I've added a new check which prevents RDDs from being directly broadcasted.

I should probably add these checks to PySpark, too.  I'm not actually sure what happens if you try to do these invalid things in PySpark, so I should probably try them first and add their errors / stacktraces to the JIRA so that it's easier for me / the support team to pattern-match to this issue.

SparkQA commented:
[Test build #25036 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25036/consoleFull) for   PR 3884 at commit [`57cc8a1`](https://github.com/apache/spark/commit/57cc8a11266770e145c7ca810bec3b95aeefabb3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25036/
Test FAILed.

JoshRosen commented:
Haha, the `org.apache.spark.broadcast.BroadcastSuite.Using broadcast after destroy prints callsite` test actually broadcasts an RDD (which is invalid), which is what caused that test failure.  I'll fix this up in my next commit.

rxin commented on core/src/main/scala/org/apache/spark/rdd/RDD.scala:
```diff
@@ -76,10 +76,22 @@ import org.apache.spark.util.random.{BernoulliSampler, PoissonSampler, Bernoulli
+  private def sc: SparkContext = {
+    if (_sc == null) {
+      throw new SparkException(
+        "Can only define RDDs and perform actions on the driver, not in tasks (see SPARK-5063)")
```
Pointing to a JIRA ticket might not be the most friendly way for users. Maybe make it more verbose and explain it in one or two lines?

JoshRosen replied:
Sure.  How about this:

> RDD transformations and actions can only be invoked by the driver, not inside of other transformations; for example, `rdd1.map(x => rdd2.values.count() * x)` is invalid because the `values` transformation and `count` action cannot be performed inside of the `rdd1.map` transformation.  For more information, see SPARK-5063.

Kind of verbose, but I think an example might be the clearest way to explain this, esp. to someone unfamiliar with the terminology.

It might be nice to keep the JIRA reference since it will make the exception easier to search for (I'm kind of inspired by React.js's error messages, which include URL-shortened links to the documentation).

rxin replied:
Looks good!

rxin commented on core/src/test/scala/org/apache/spark/rdd/RDDSuite.scala:
```diff
@@ -897,4 +897,23 @@ class RDDSuite extends FunSuite with SharedSparkContext {
     }
   }
+
+  test("Nested RDDs are not supported (SPARK-5063)") {
```
a nit pick: i don't think we have a standard, but so far test case names start with lower case.

JoshRosen replied:
It varies from suite-to-suite; most start with lowercase because they start with method names.  If you look at BroadcastSuite, though, most use uppercase style like I've done here.

JoshRosen replied:
A quick `git grep` suggests that every suite uses its own style and that there's not an obvious dominant style.  I'll just change these tests to the lowercase convention to match RDDSuite, but leave the BroadcastSuite ones as-is.

JoshRosen committed:
Add more descriptive error message.

JoshRosen committed:
Guard against calling methods on stopped SparkContexts.

SparkQA commented:
[Test build #25084 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25084/consoleFull) for   PR 3884 at commit [`99cc09f`](https://github.com/apache/spark/commit/99cc09f6996706f5d067d878d486d8f5dc2c31f7).
 * This patch merges cleanly.

JoshRosen commented:
I've added some additional tests to prevent users from calling methods on a stopped SparkContext, since this usually resulted in confusing NullPointerExceptions.

JoshRosen committed:
Fix BroadcastSuite test which broadcasted an RDD

SparkQA commented:
[Test build #25085 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25085/consoleFull) for   PR 3884 at commit [`b39e041`](https://github.com/apache/spark/commit/b39e04172d46b036c467b1650f7c27f799bfdfc0).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25084 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25084/consoleFull) for   PR 3884 at commit [`99cc09f`](https://github.com/apache/spark/commit/99cc09f6996706f5d067d878d486d8f5dc2c31f7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25084/
Test FAILed.

SparkQA commented:
[Test build #25085 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25085/consoleFull) for   PR 3884 at commit [`b39e041`](https://github.com/apache/spark/commit/b39e04172d46b036c467b1650f7c27f799bfdfc0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25085/
Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #25114 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25114/consoleFull) for   PR 3884 at commit [`b39e041`](https://github.com/apache/spark/commit/b39e04172d46b036c467b1650f7c27f799bfdfc0).
 * This patch merges cleanly.

JoshRosen commented:
Any opinions on the `assertNotStopped()` checks here?  I'd like to backport this patch to other branches since I think it's a huge usability improvement.  If there are any changes here that you think might break user programs that used to work, then I'll remove them and re-add them in a separate PR.

(Note: I still need to do the PySpark half of the "nested RDDs" and "nested actions" checks)

rxin commented:
Maybe IllegalStateException?

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into SPARK-5063

JoshRosen committed:
SparkException -> IllegalStateException

JoshRosen committed:
Add improved error messages to PySpark.

JoshRosen commented:
Alright, I've updated this to use IllegalStateException when methods are called on a stopped SparkContext.  I've also added some more helpful error messages to PySpark when users attempt to mis-use SparkContext or RDDs from actions, transformations, or broadcast variables.

I plan to merge this into master, then backport a smaller patch which excludes most of the `assertNotStopped()` calls.

SparkQA commented:
[Test build #25426 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25426/consoleFull) for   PR 3884 at commit [`9f6a0b8`](https://github.com/apache/spark/commit/9f6a0b8d3501b2872e75f1eff0bf1e4b765183e0).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25426 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25426/consoleFull) for   PR 3884 at commit [`9f6a0b8`](https://github.com/apache/spark/commit/9f6a0b8d3501b2872e75f1eff0bf1e4b765183e0).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25426/
Test FAILed.

JoshRosen committed:
Fix Python line length issues.

SparkQA commented:
[Test build #25433 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25433/consoleFull) for   PR 3884 at commit [`6ef68d0`](https://github.com/apache/spark/commit/6ef68d050ca5ef52b25f10537c9e0ac44562ebc0).
 * This patch merges cleanly.

SparkQA commented:
**[Test build #25433 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25433/consoleFull)**     for PR 3884 at commit [`6ef68d0`](https://github.com/apache/spark/commit/6ef68d050ca5ef52b25f10537c9e0ac44562ebc0)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25433/
Test FAILed.

JoshRosen committed:
IllegalStateException fix

SparkQA commented:
[Test build #25441 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25441/consoleFull) for   PR 3884 at commit [`8cff41a`](https://github.com/apache/spark/commit/8cff41aa0b2a22573e61e413e972c727eb6782a8).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25441 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25441/consoleFull) for   PR 3884 at commit [`8cff41a`](https://github.com/apache/spark/commit/8cff41aa0b2a22573e61e413e972c727eb6782a8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25441/
Test PASSed.

JoshRosen commented:
@andrewor14 @pwendell @tdas How do you feel about committing this patch, as-is, for 1.2.1?  I think it could be a huge support burden reducer / usability improver for many users, since a lot of these issues are really hard to debug.

If you'd like, I can grab a copy of branch-1.2 and manually check that all of the `assertNotStopped` methods threw errors (just to make sure that we're not missing any corner-cases where we change behavior for something that used to work).

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -526,6 +534,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    * the argument to avoid this.
    */
   def parallelize[T: ClassTag](seq: Seq[T], numSlices: Int = defaultParallelism): RDD[T] = {
```
In 1.2, calling this when SparkContext was stopped would throw a NullPointerException:

```
scala> sc.parallelize(1 to 100)
java.lang.NullPointerException
	at org.apache.spark.SparkContext.defaultParallelism(SparkContext.scala:1461)
	at org.apache.spark.SparkContext.parallelize$default$2(SparkContext.scala:521)
	at $iwC$$iwC$$iwC$$iwC.<init>(<console>:13)
	at $iwC$$iwC$$iwC.<init>(<console>:18)
	at $iwC$$iwC.<init>(<console>:20)
	at $iwC.<init>(<console>:22)
	at <init>(<console>:24)
	at .<init>(<console>:28)
	at .<clinit>(<console>)
	at .<init>(<console>:7)
	at .<clinit>(<console>)
	at $print(<console>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.spark.repl.SparkIMain$ReadEvalPrint.call(SparkIMain.scala:852)
	at org.apache.spark.repl.SparkIMain$Request.loadAndRun(SparkIMain.scala:1125)
	at org.apache.spark.repl.SparkIMain.loadAndRunReq$1(SparkIMain.scala:674)
	at org.apache.spark.repl.SparkIMain.interpret(SparkIMain.scala:705)
	at org.apache.spark.repl.SparkIMain.interpret(SparkIMain.scala:669)
	at org.apache.spark.repl.SparkILoop.reallyInterpret$1(SparkILoop.scala:828)
	at org.apache.spark.repl.SparkILoop.interpretStartingWith(SparkILoop.scala:873)
	at org.apache.spark.repl.SparkILoop.command(SparkILoop.scala:785)
	at org.apache.spark.repl.SparkILoop.processLine$1(SparkILoop.scala:628)
	at org.apache.spark.repl.SparkILoop.innerLoop$1(SparkILoop.scala:636)
	at org.apache.spark.repl.SparkILoop.loop(SparkILoop.scala:641)
	at org.apache.spark.repl.SparkILoop$$anonfun$process$1.apply$mcZ$sp(SparkILoop.scala:968)
	at org.apache.spark.repl.SparkILoop$$anonfun$process$1.apply(SparkILoop.scala:916)
	at org.apache.spark.repl.SparkILoop$$anonfun$process$1.apply(SparkILoop.scala:916)
	at scala.tools.nsc.util.ScalaClassLoader$.savingContextLoader(ScalaClassLoader.scala:135)
	at org.apache.spark.repl.SparkILoop.process(SparkILoop.scala:916)
	at org.apache.spark.repl.SparkILoop.process(SparkILoop.scala:1011)
	at org.apache.spark.repl.Main$.main(Main.scala:31)
	at org.apache.spark.repl.Main.main(Main.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.spark.deploy.SparkSubmit$.launch(SparkSubmit.scala:358)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:75)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
```

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -550,6 +560,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    * Hadoop-supported file system URI, and return it as an RDD of Strings.
    */
   def textFile(path: String, minPartitions: Int = defaultMinPartitions): RDD[String] = {
+    assertNotStopped()
```
Same for textFile:

```
scala> sc.textFile("/usr/share/dict/words")
java.lang.NullPointerException
	at org.apache.spark.SparkContext.defaultParallelism(SparkContext.scala:1461)
	at org.apache.spark.SparkContext.defaultMinPartitions(SparkContext.scala:1468)
	at org.apache.spark.SparkContext.textFile$default$2(SparkContext.scala:545)
```

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -891,14 +913,22 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
 
   /** Get an RDD that has no partitions or elements. */
-  def emptyRDD[T: ClassTag] = new EmptyRDD[T](this)
+  def emptyRDD[T: ClassTag] = {
```
This did _not_ cause an exception in 1.2, so I'll remove this `assertNotStopped` call.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -891,14 +913,22 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
 
   /** Build the union of a list of RDDs. */
-  def union[T: ClassTag](rdds: Seq[RDD[T]]): RDD[T] = new UnionRDD(this, rdds)
+  def union[T: ClassTag](rdds: Seq[RDD[T]]): RDD[T] = {
```
Same here; instantiating new UnionRDDs doesn't cause an error if SC is stopped.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -906,8 +936,10 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    * Create an [[org.apache.spark.Accumulator]] variable of a given type, which tasks can "add"
    * values to using the `+=` method. Only the driver can access the accumulator's `value`.
    */
-  def accumulator[T](initialValue: T)(implicit param: AccumulatorParam[T]) =
```
Same for these accumulator methods, so I'll revert these changes.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -955,6 +993,11 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    * The variable will be sent to each cluster only once.
    */
   def broadcast[T: ClassTag](value: T): Broadcast[T] = {
```
Broadcast, on the other hand, throws a NPE:

```
scala> sc.broadcast(0)
java.lang.NullPointerException
	at org.apache.spark.broadcast.TorrentBroadcast.<init>(TorrentBroadcast.scala:79)
	at org.apache.spark.broadcast.TorrentBroadcastFactory.newBroadcast(TorrentBroadcastFactory.scala:34)
	at org.apache.spark.broadcast.TorrentBroadcastFactory.newBroadcast(TorrentBroadcastFactory.scala:29)
	at org.apache.spark.broadcast.BroadcastManager.newBroadcast(BroadcastManager.scala:62)
	at org.apache.spark.SparkContext.broadcast(SparkContext.scala:951)
	at $iwC$$iwC$$iwC$$iwC.<init>(<console>:13)
	at $iwC$$iwC$$iwC.<init>(<console>:18)
	at $iwC$$iwC.<init>(<console>:20)
	at $iwC.<init>(<console>:22)
	at <init>(<console>:24)
	at .<init>(<console>:28)
	at .<clinit>(<console>)
	at .<init>(<console>:7)
	at .<clinit>(<console>)
	at $print(<console>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
```

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -969,6 +1012,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    * use `SparkFiles.get(fileName)` to find its download location.
    */
   def addFile(path: String) {
```
This throws an NPE:

```
scala> sc.addFile("/usr/share/dict/words")
java.lang.NullPointerException
	at org.apache.spark.SparkFiles$.getRootDirectory(SparkFiles.scala:37)
	at org.apache.spark.SparkContext.addFile(SparkContext.scala:975)
	at $iwC$$iwC$$iwC$$iwC.<init>(<console>:13)
	at $iwC$$iwC$$iwC.<init>(<console>:18)
	at $iwC$$iwC.<init>(<console>:20)
	at $iwC.<init>(<console>:22)
	at <init>(<console>:24)
	at .<init>(<console>:28)
	at .<clinit>(<console>)
	at .<init>(<console>:7)
	at .<clinit>(<console>)
	at $print(<console>)
```

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -992,6 +1036,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    */
   @DeveloperApi
   def addSparkListener(listener: SparkListener) {
+    assertNotStopped()
```
AddSparkListener technically works, so I'll remove this, too.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1002,6 +1047,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    */
   @DeveloperApi
   override def requestExecutors(numAdditionalExecutors: Int): Boolean = {
+    assertNotStopped()
```
This is a toss-up, since I'd expect to throw some sort of error after the scheduler backend is stopped, but there are many cases where it's a no-op and doesn't throw an error.  Therefore, I'll remove this, too.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1047,6 +1097,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    * memory available for caching.
    */
   def getExecutorMemoryStatus: Map[String, (Long, Long)] = {
```
This throws an error, so I'll keep it:

```
scala> sc.getExecutorMemoryStatus
org.apache.spark.SparkException: Error sending message as actor is null [message = GetMemoryStatus]
	at org.apache.spark.util.AkkaUtils$.askWithReply(AkkaUtils.scala:178)
	at org.apache.spark.storage.BlockManagerMaster.askDriverWithReply(BlockManagerMaster.scala:221)
	at org.apache.spark.storage.BlockManagerMaster.getMemoryStatus(BlockManagerMaster.scala:148)
	at org.apache.spark.SparkContext.getExecutorMemoryStatus(SparkContext.scala:1039)
	at $iwC$$iwC$$iwC$$iwC.<init>(<console>:13)
	at $iwC$$iwC$$iwC.<init>(<console>:18)
	at $iwC$$iwC.<init>(<console>:20)
	at $iwC.<init>(<console>:22)
	at <init>(<console>:24)
	at .<init>(<console>:28)
	at .<clinit>(<console>)
	at .<init>(<console>:7)
	at .<clinit>(<console>)
	at $print(<console>)
```

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1059,6 +1110,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    */
   @DeveloperApi
   def getRDDStorageInfo: Array[RDDInfo] = {
+    assertNotStopped()
```
Same here:

```
scala> sc.getRDDStorageInfo
org.apache.spark.SparkException: Error sending message as actor is null [message = GetStorageStatus]
	at org.apache.spark.util.AkkaUtils$.askWithReply(AkkaUtils.scala:178)
	at org.apache.spark.storage.BlockManagerMaster.askDriverWithReply(BlockManagerMaster.scala:221)
	at org.apache.spark.storage.BlockManagerMaster.getStorageStatus(BlockManagerMaster.scala:152)
	at org.apache.spark.SparkContext.getExecutorStorageStatus(SparkContext.scala:1068)
	at org.apache.spark.SparkContext.getRDDStorageInfo(SparkContext.scala:1052)
	at $iwC$$iwC$$iwC$$iwC.<init>(<console>:13)
	at $iwC$$iwC$$iwC.<init>(<console>:18)
	at $iwC$$iwC.<init>(<console>:20)
	at $iwC.<init>(<console>:22)
	at <init>(<console>:24)
	at .<init>(<console>:28)
	at .<clinit>(<console>)
	at .<init>(<console>:7)
	at .<clinit>(<console>)
	at $print(<console>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(N
```

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1068,7 +1120,10 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    * Returns an immutable map of RDDs that have marked themselves as persistent via cache() call.
    * Note that this does not necessarily mean the caching or computation was successful.
    */
-  def getPersistentRDDs: Map[Int, RDD[_]] = persistentRdds.toMap
```
This is safe, so I'll revert this error-checking.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1111,6 +1170,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    */
   @deprecated("adding files no longer creates local copies that need to be deleted", "1.0.0")
   def clearFiles() {
+    assertNotStopped()
```
I'll remove this.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1146,6 +1206,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    * filesystems), an HTTP, HTTPS or FTP URI, or local:/path for a file on every worker node.
    */
   def addJar(path: String) {
+    assertNotStopped()
```
This is another one of those toss-up cases: it works some of the time, so I'll remove this check for conservatism's sake.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1199,6 +1260,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    */
   @deprecated("adding jars no longer creates local copies that need to be deleted", "1.0.0")
   def clearJars() {
+    assertNotStopped()
```
I'll revert this.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1458,6 +1522,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    * be a HDFS path if running on a cluster.
    */
   def setCheckpointDir(directory: String) {
+    assertNotStopped()
```
This actually works, so I'm removing this check.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1466,17 +1531,29 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
 
   /** Default level of parallelism to use when not given by user (e.g. parallelize and makeRDD). */
-  def defaultParallelism: Int = taskScheduler.defaultParallelism
+  def defaultParallelism: Int = {
```
This throws an exception because `taskScheduler` is null:

```
scala> sc.defaultParallelism
java.lang.NullPointerException
	at org.apache.spark.SparkContext.defaultParallelism(SparkContext.scala:1461)
	at $iwC$$iwC$$iwC$$iwC.<init>(<console>:13)
	at $iwC$$iwC$$iwC.<init>(<console>:18)
	at $iwC$$iwC.<init>(<console>:20)
	at $iwC.<init>(<console>:22)
	at <init>(<console>:24)
	at .<init>(<console>:28)
	at .<clinit>(<console>)
	at .<init>(<console>:7)
	at .<clinit>(<console>)
	at $print(<console>)
```

JoshRosen committed:
Remove assertNotStopped() calls for methods that were sometimes safe to call on stopped SC's in Spark 1.2

JoshRosen committed:
Revert two unintentional formatting changes

JoshRosen committed:
Fix test to reflect 1.2.1 compatibility

JoshRosen commented:
I audited the uses of `assertNotStopped` and removed a bunch of calls in methods that sometimes didn't throw exceptions on Spark 1.2.0.  Pending Jenkins, I'm planning to commit this slightly smaller patch to branch-1.2 for inclusion in 1.2.1.

SparkQA commented:
[Test build #25928 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25928/consoleFull) for   PR 3884 at commit [`2d0d7f7`](https://github.com/apache/spark/commit/2d0d7f78535a193e96309c81b3a6a5fded71fe48).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25928 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25928/consoleFull) for   PR 3884 at commit [`2d0d7f7`](https://github.com/apache/spark/commit/2d0d7f78535a193e96309c81b3a6a5fded71fe48).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25928/
Test PASSed.

JoshRosen commented:
Alright, this is ready for a final review + commit if anyone wants to take a look.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -955,6 +977,11 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    */
   def broadcast[T: ClassTag](value: T): Broadcast[T] = {
+    assertNotStopped()
+    if (classOf[RDD[_]].isAssignableFrom(classTag[T].runtimeClass)) {
```
Actually, maybe this check should go somewhere else, since I think that it might technically have been safe to _create_ a broadcast variable with an RDD, even though doing anything with it would trigger errors.

JoshRosen replied:
I've changed this in my latest patch; we log a warning here and any errors are caught by the more general "display an error about RDD nesting if the `sc` field is null" check.

JoshRosen commented on core/src/main/scala/org/apache/spark/rdd/RDD.scala:
```diff
@@ -76,10 +76,25 @@ import org.apache.spark.util.random.{BernoulliSampler, PoissonSampler, Bernoulli
     @transient private var deps: Seq[Dependency[_]]
   ) extends Serializable with Logging {
 
+  if (classOf[RDD[_]].isAssignableFrom(elementClassTag.runtimeClass)) {
```
Similarly, this should perhaps be a warning instead of an exception in order to avoid any possibility of breaking odd corner-case 1.2.1 apps.  I'll change this to a warning and leave the `sc` getter as an exception.

JoshRosen committed:
Convert two exceptions into warnings in order to avoid breaking user programs in some edge-cases.

SparkQA commented:
[Test build #26026 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26026/consoleFull) for   PR 3884 at commit [`a943e00`](https://github.com/apache/spark/commit/a943e00fd76d1b84a598fa449b5abd99074c2c62).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26026 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26026/consoleFull) for   PR 3884 at commit [`a943e00`](https://github.com/apache/spark/commit/a943e00fd76d1b84a598fa449b5abd99074c2c62).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26026/
Test PASSed.

vanzin commented on python/pyspark/rdd.py:
```diff
@@ -141,6 +141,17 @@ def id(self):
+        # This method is called when attempting to pickle an RDD, which is always an error:
+        raise Exception(
+            "It appears that you are attempting to broadcast an RDD or reference an RDD from an "
+            "action or transforamtion. RDD transformations and actions can only be invoked by the "
```
typo: "transforamtion"

vanzin commented:
Scala changes look ok to me; I'm not super familiar with the pyspark internals, but the check on `rdd.py` surprised me because I thought RDDs were actually serialized, at least on the Scala side.

JoshRosen commented:
@vanzin Thanks for looking this over.  The Python `RDD` objects themselves are never actually serialized and are used internally in a way that's slightly different than in Scala/Java Spark.  In the existing code, any attempt to serialize instances of those Python classes throws an exception in the `__getnewargs__` method, which is why I was able to add new exceptions there.

I'm going to fix the spelling error, take one final look over this, and commit it so we can get it into the first 1.2.1 RC.  I saw a couple of mailing list questions yesterday that could have been prevented by this patch, which illustrates why I really want to get this into our next maintenance release.

JoshRosen committed:
Fix spelling typo

SparkQA commented:
[Test build #26037 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26037/consoleFull) for   PR 3884 at commit [`a38774b`](https://github.com/apache/spark/commit/a38774b8892a85184520078a2187e9ce2a190038).
 * This patch merges cleanly.

JoshRosen commented:
Alright, I've merged this into `master` (1.3.0) and `branch-1.2` (1.2.1).

SparkQA commented:
[Test build #26037 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26037/consoleFull) for   PR 3884 at commit [`a38774b`](https://github.com/apache/spark/commit/a38774b8892a85184520078a2187e9ce2a190038).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26037/
Test PASSed.

nchammas commented:
Thank you @JoshRosen for working on usability issues like this.

kschandramouli commented:
Hi Guys, I ran into this issue in my ETL Script. Getting this:

An error was encountered:
Could not serialize object: Exception: It appears that you are attempting to reference SparkContext from a broadcast variable, action, or transformation. SparkContext can only be used on the driver, not in code that it run on workers. For more information, see SPARK-5063.

Error Occurs when I try to access this code:

 # Loads parquet file located in AWS S3 into RDD Data Frame
    parquetFile = sqlContext.read.parquet("s3://list-etl")

Can someone please advise? Please find the entire piece of code below:

import sys
import json
from awsglue.transforms import *
from awsglue.utils import getResolvedOptions
from pyspark.context import SparkContext
from awsglue.context import GlueContext
from awsglue.dynamicframe import DynamicFrame
from awsglue.job import Job

from pyspark.sql.functions import lit
#from urllib.parse as urlparse, parse_qs
#from urllib import parse as urlparse
#from urllib import parse.parse_qs as parse_qs
#from urllib.parse import urlparse, parse_qs
from Crypto.Cipher import AES
from base64 import *
from pyspark.sql import SQLContext
from pyspark.sql.types import *
from urllib.parse import urlparse, parse_qs
from Crypto.Cipher import AES
from base64 import *

from pyspark.sql import SparkSession
from pyspark.sql.functions import udf
from pyspark.sql.functions import broadcast
from pyspark.sql.types import StringType
import logging


#try:
#    from pyspark import SQLContext
#    print ("Successfully imported SQLContext Modules -- 'SQLContext'")
#except ImportError as e:
#    print ("Can not import Spark Modules", e)
#    sys.exit(1)

args = getResolvedOptions(sys.argv, ['TempDir',
                                     'JOB_NAME',
                                     'databaseName',
                                     's3bucketETL',
                                     'arcGIStokenKey',
                                     'tableName'])

def map_function(dynamicRecord):
    URL = dynamicRecord["target_request"]
    
    # need this step to remove the leading POST/GET and trailing HTTP version
    splitURL = URL.split(" ")
    
    # parse only the URL (second item)
    parsedURL = urlparse(splitURL[1])
    dynamicRecord["url_path"] = str(parsedURL.path)

    # parse the query string component
    parsedQuery = parse_qs(parsedURL.query)
    dynamicRecord["url_query"] = str(json.dumps(parsedQuery))
    
    # decode the token, if present
    if 'token' in parsedQuery:
        token_input = str(parsedQuery['token'])
        raw_token = token_input[3:-2] 
        dynamicRecord["url_token"] = str(raw_token)
        
        iv = '\x01'+ '\x02'+ '\x03'+ '\x04'+ '\x05'+ '\x06'+ '\x07'+ '\x08'+ '\x09'+ '\x0A'+ '\x0B'+ '\x0C'+ '\x0D'+ '\x0E'+ '\x0F' + '\x10'

        # pad the token with extra '=' to make sure that it is divisible by 16, and therefore base64 
        pad = (len(raw_token))%16
        raw_token += "="*pad
        
        try:
            raw_token = b64decode(raw_token, '-_')
        except:
            dynamicRecord["user_id"] = str('raw_token_decode_error')
        else:
            mode = AES.MODE_CBC
            # Make sure your key is 16 chars. if it's shorter pad it with 0s
            # longer then trim it at 16.
            key = args['arcGIStokenKey'] # padding if needed +('\x00'*6)
            cipher = AES.new(key, mode, iv)
            
            try:
                # decrypt the token
                binary_decrypted_token = cipher.decrypt(raw_token)
            except:
                dynamicRecord["user_id"] = str('raw_token_decrypt_error')
            else:
                # decode it from binary for easier visualisation and text processing
                decoded_token = binary_decrypted_token.decode('utf-8', 'ignore')

                # split the token into its components (user ID, token expiry, , referer)
                split_token = decoded_token.split(":")

                dynamicRecord["user_id"] = str(split_token[0])

    # convert the log timestamp_response to a friendly format without the timezone
    timestamp_response = dynamicRecord["timestamp_response"]
    date_response = timestamp_response [:10]
    time_response = timestamp_response [11:25]    
    new_timestamp = date_response+" "+time_response    
    dynamicRecord["timestamp_response"] = new_timestamp

    # convert the log timestamp_request to a friendly format without the timezone
    timestamp_request = dynamicRecord["timestamp_request"]
    date_request = timestamp_request [:10]
    time_request = timestamp_request [11:25]    
    new_timestamp = date_request+" "+time_request    
    dynamicRecord["timestamp_request"] = new_timestamp

    #######################################
    # Ascertain the SLIP service elements #
    #######################################

    # split the URL into qualifiers
    parsedUrlQualifiers = parsedURL.path.split("/")
    # remove any empty strings
    parsedUrlQualifiers = [i for i in parsedUrlQualifiers if i]

    # find the base index for the services component. if it's not found, return the record without the asset information
    try:
        servicesIndex = parsedUrlQualifiers.index('services')
    except:
        try:
            servicesIndex = parsedUrlQualifiers.index('datadownloads')
        except:
            return dynamicRecord

    # find the base index for the server component if it's for "services"
    if parsedUrlQualifiers[servicesIndex] == 'services':
        servicesURL = True
        try:
            serverIndex = parsedUrlQualifiers.index('MapServer')
        except:
            try:
                serverIndex = parsedUrlQualifiers.index('FeatureServer')
            except:
                try:
                    serverIndex = parsedUrlQualifiers.index('ImageServer')
                except:
                    return dynamicRecord
        dynamicRecord["service_type"] = str(get_service_type(parsedUrlQualifiers, serverIndex))
    else:
        servicesURL = False
        dynamicRecord["service_type"] = str('Data Download')

    assetGroup = get_asset_group(parsedUrlQualifiers, servicesIndex)
    dynamicRecord["asset_group"] = str(assetGroup)
    dynamicRecord["service_name"] = str(get_service_name(parsedUrlQualifiers, servicesIndex))
    dynamicRecord["service_class"] = str(get_service_class(assetGroup))
    
    if servicesURL:
        requestType = get_request_type(parsedQuery, parsedUrlQualifiers, serverIndex)
        dynamicRecord["request_type"] = str(requestType)
        layerList = get_layer_list(requestType, parsedQuery, parsedUrlQualifiers, serverIndex)
        #layerNames = get_layer_names_services()
        #dynamicRecord["layer_names"] = str(layerList)  

        #slip_layer_info_sql= get_slip_layer_info("slip_reference_database","slip_reference_etl_slip_layer_list_etl",str(dynamicRecord["layer_names"]),str(dynamicRecord["service_type"]),str(dynamicRecord["service_name"]))
    
        slip_layer_info_sql=[]
        slip_layer_info_sql= get_slip_layer_info(str(layerList),dynamicRecord["service_type"],dynamicRecord["service_name"])

        log.info("log4jLogger: slip_layer_info_sql" + slip_layer_info_sql[0])
        log.info("log4jLogger: slip_layer_info_sql" + slip_layer_info_sql[1])
        log.info("log4jLogger: slip_layer_info_sql" + slip_layer_info_sql[2])

        dynamicRecord["slip_layer_ids"]  = str(slip_layer_info_sql[0])
        dynamicRecord["layer_names"] = str(slip_layer_info_sql[1])
        dynamicRecord["slip_organisation_id"] = str(slip_layer_info_sql[2])
    else:
        layerNames = get_layer_name_datadownload(parsedUrlQualifiers)
        dynamicRecord["layer_names"] = str(layerNames)

    if not servicesURL:
        dynamicRecord["datadownload_ext"] = str(get_download_format(parsedUrlQualifiers))

    return dynamicRecord

#def get_slip_layer_info(layerListDBName,layerListTableName,layerID, serviceType, serviceName):
def get_slip_layer_info(sqlContext, layerID, serviceType, serviceName):
    layer_list=[]
    #sqlContext = SQLContext(sparkContext=sc)
   
    # Loads parquet file located in AWS S3 into RDD Data Frame
    parquetFile = sqlContext.read.parquet("s3://slip-layer-list-etl")

    # Stores the DataFrame into an "in-memory temporary table"
    parquetFile.registerTempTable("parquetFile")

    # Run standard SQL queries against temporary table
    #reference_all_sql = sqlContext.sql("SELECT * FROM parquetFile")

    # Use standard SQL to filter
    reference_filtered_sql = sqlContext.sql("SELECT distinct layer_id, layer_name, slip_layer_id FROM parquetFile WHERE layer_id LIKE '%" + layerID + "%' and service_type LIKE '%" + service_type + "%' and service_name LIKE '%" + serviceName + "%'")
    #reference_filtered_sql = sqlContext.sql("select distinct layer_id, layer_name, slip_layer_id FROM parquetFile where layer_id LIKE '%10%' and service_type LIKE '%MapServer%' and service_name LIKE '%NLRP_Query_Services%'")

    layer_id_val = reference_filtered_sql.rdd.map(lambda p: p.layer_id)
    for layer_id in layer_id_val.collect():
        layer_list.append(str(layer_id))

    layer_names_val = reference_filtered_sql.rdd.map(lambda p: p.layer_name)
    for layer_name in layer_names_val.collect():
        layer_list.append(str(layer_name))

    slip_organisation_id_val = reference_filtered_sql.rdd.map(lambda p: p.slip_layer_id)
    for slip_organisation_id in slip_organisation_id_val.collect():
        layer_list.append(str(slip_organisation_id))

    print("layer_list[0]" + layer_list[0])
    print("layer_list[1]" + layer_list[1])
    print("layer_list[2]" + layer_list[2])
    return layer_list

def get_asset_group(urlList, startingIndex):
    assetGroup = urlList[startingIndex + 1]
    return assetGroup

def get_service_name(urlList, startingIndex):
    serviceName = urlList[startingIndex + 2]
    serviceName = serviceName.split("_FS", 1)[0]
    serviceName = serviceName.split("_WFS", 1)[0]
    return serviceName

def get_service_type(urlList, startingIndex):
    try:
        serviceType = urlList[startingIndex + 1]
    except:
        serviceType = urlList[startingIndex]
    else:
        if serviceType == '':
            serviceType = 'MapServer'
        elif serviceType == 'export':
            serviceType = 'MapServer'
        elif serviceType == 'identify':
            serviceType = 'MapServer'
        elif serviceType == 'layers':
            serviceType = 'MapServer'
        elif serviceType == 'legend':
            serviceType = 'MapServer'
        elif serviceType == 'WMTS':
            pass
        else:
            try:
                queryTest = urlList[startingIndex + 2]
            except:
                serviceType = urlList[startingIndex + 1]
            else:
                serviceType = urlList[startingIndex]

    return serviceType

def get_service_class(assetGroupString):
    serviceClasses = ['Public_Secure', 'Public', 'Subscription', 'Restricted', 'Government_Only']
    for serviceClass in serviceClasses:
        if serviceClass in assetGroupString:
            return serviceClass

    return ''

def get_request_type(queryDict, urlList, startingIndex):
    if 'request' in queryDict:
        rawRequestType = str(queryDict['request'])
        requestType = str(rawRequestType[3:-2])
    elif urlList[startingIndex] == 'MapServer':
        try:
            testEndUrl = urlList[startingIndex + 2]
        except:
            try:
                testEndUrl = urlList[startingIndex + 1]
            except:
                requestType = 'Service Capabilities'
            else:
                if testEndUrl == 'export':
                    requestType = 'Export Map'
                elif testEndUrl == 'identify':
                    requestType = 'Identify Layers'
                elif testEndUrl == 'layers':
                    requestType = 'Service Layer Listing'
                elif testEndUrl == 'legend':
                    requestType = 'Service Layer Legends'
        else:
            if testEndUrl == 'query':
                requestType = 'Query Layer'
            elif urlList[startingIndex + 1] == 'WMTS':
                requestType = 'WMTS'
    elif urlList[startingIndex] == 'FeatureServer':
        try:
            testEndUrl = urlList[startingIndex + 2]
        except:
            try:
                testEndUrl = urlList[startingIndex + 1]
            except:
                requestType = 'Service Capabilities'
        else:
            if testEndUrl.isdigit():
                requestType = 'Get Single Feature'
            elif testEndUrl == 'query':
                requestType = 'Query Layer'
    elif urlList[startingIndex] == 'ImageServer':
        requestType = ''
    else:
        requestType = ''

    return requestType

def get_layer_list(rType,queryDict,urlList,serverIndex):
    if rType == "GetMap":
        if "layers" in queryDict:
            layerList = queryDict["layers"]
    elif rType == "GetFeatureInfo":
        if "query_layers" in queryDict:
            layerList = queryDict["query_layers"]
    elif rType == "GetLegendGraphic":
        if "layer" in queryDict:
            layerList = queryDict["layer"]
    elif rType == "Service Capabilities":
        layerList = ""
    elif rType == "Export Map":
        if "layers" in queryDict:
            if str(queryDict["layers"][0][:5]) == "show:":
                layerList = str(queryDict["layers"][0][5:]).split(",")
                try:
                    layerList.remove("-1")
                except:
                    pass
                else:
                    layerList = [i for i in layerList if i >= "0"]
            else:
                layerList = ""
    elif rType == "Identify Layers":
        if "layers" in queryDict:
            if str(queryDict["layers"][0][:8]) == "visible:":
                layerList = str(queryDict["layers"][0][8:]).split(",")
                try:
                    layerList.remove("-1")
                except:
                    pass
                else:
                    layerList = [i for i in layerList if i >= "0"]
            else:
                layerList = ""
    elif rType == "Query Layer":
        try:
            layerList = urlList[serverIndex + 1]
        except:
            layerList = ""
    elif rType == "DescribeFeatureType":
        if "typeName" in queryDict:
            typeName = queryDict["typeName"]
            layerList = []
            layerIds = str(queryDict["typeName"][0]).split(",")
            for layer in layerIds:
                colonIndex = layer.find(":")
                if colonIndex == '-1':
                    layerList.append(layer)
                else:
                    layerList.append(layer[(colonIndex + 1):])
        else:
            layerList = ""
    elif rType == "GetFeature":
        if "typeName" in queryDict:
            typeName = queryDict["typeName"]
            layerList = []
            layerIds = str(queryDict["typeName"][0]).split(",")
            for layer in layerIds:
                colonIndex = layer.find(":")
                if colonIndex == '-1':
                    layerList.append(layer)
                else:
                    layerList.append(layer[(colonIndex + 1):])
        else:
            layerList = ""
    elif rType == "Get Single Feature":
        try:
            layerList = urlList[serverIndex + 1]
        except:
            layerList = ""
    else:
        layerList = ""

    return layerList

def get_layer_name_datadownload(urlList):
    layerName = urlList[-2]

    return layerName

def get_download_format(urlList):
    downloadedFile = urlList[-1]
    downloadedFileExt = '.' + downloadedFile.split(".", 1)[1]

    return downloadedFileExt

sc = SparkContext()
glueContext = GlueContext(sc)
spark = glueContext.spark_session
job = Job(glueContext)
job.init(args['JOB_NAME'], args)

#log4jLogger = spark.sparkContext._jvm.org.apache.log4j
#log = log4jLogger.LogManager.getLogger(__name__)

## @type: DataSource
## @args: [database = "slip-landgate-database", table_name = "slip_landgate_alb_ap_southeast_2", transformation_ctx = "datasource0"]
## @return: datasource0
## @inputs: []
datasource0 = glueContext.create_dynamic_frame.from_catalog(database = args['databaseName'], table_name = args['tableName'], transformation_ctx = "datasource0")

# convert to Spark DataFrame to add nine new columns; url_path, url_query, url_token, user_id etc., all strings
spark_df = datasource0.toDF()
spark_df = spark_df.withColumn("url_path", lit(" ")).withColumn("url_query", lit(" ")).withColumn("url_token", lit(" ")).withColumn("user_id", lit(" ")).withColumn("asset_group", lit(" ")).withColumn("service_name", lit(" ")).withColumn("layer_names", lit(" ")).withColumn("slip_layer_ids", lit(" ")).withColumn("slip_organisation_id", lit(" ")).withColumn("service_type", lit(" ")).withColumn("service_class", lit(" ")).withColumn("request_type", lit(" ")).withColumn("datadownload_ext", lit(" "))

# convert back to Glue DynamicFrame for column manipulation
datasourceDyF = DynamicFrame.fromDF(spark_df, glueContext, "newcols")
convertdynamic = Map.apply(frame = datasourceDyF, f = map_function)

## @type: ApplyMapping
## @args: [mapping = [("http", "string", "http", "string"), ("timestamp_response", "string", "timestamp_response", "string"), ("gateway", "string", "gateway", "string"), ("client", "string", "client", "string"), ("target", "string", "target", "string"), ("request_processing_time", "string", "request_processing_time", "string"), ("target_processing_time", "string", "target_processing_time", "string"), ("response_processing_time", "string", "response_processing_time", "string"), ("elb_status_code", "string", "elb_status_code", "string"), ("target_status_code", "string", "target_status_code", "string"), ("received_bytes", "string", "received_bytes", "string"), ("sent_bytes", "string", "sent_bytes", "string"), ("target_request", "string", "target_request", "string"), ("user_agent", "string", "user_agent", "string"), ("ssl_cipher", "string", "ssl_cipher", "string"), ("ssl_protocol", "string", "ssl_protocol", "string"), ("target_group_arn", "string", "target_group_arn", "string"), ("trace_id", "string", "trace_id", "string"), ("domain_name", "string", "domain_name", "string"), ("chosen_cert_arn", "string", "chosen_cert_arn", "string"), ("matched_rule_priority", "string", "matched_rule_priority", "string"), ("timestamp_request", "string", "timestamp_request", "string"), ("actions_executed", "string", "actions_executed", "string"), ("redirect_url", "string", "redirect_url", "string"), ("error_reason", "string", "error_reason", "string"), ("partition_0", "string", "partition_0", "string"), ("partition_1", "string", "partition_1", "string"), ("partition_2", "string", "partition_2", "string"), ("partition_3", "string", "partition_3", "string"), ("partition_4", "string", "partition_4", "string"), ("partition_5", "string", "partition_5", "string"), ("partition_6", "string", "partition_6", "string")], transformation_ctx = "applymapping1"]
## @return: applymapping1
## @inputs: [frame = datasource0]
applymapping1 = ApplyMapping.apply(frame = convertdynamic, mappings = [("http", "string", "http", "string"), ("timestamp_response", "string", "timestamp_response", "timestamp"), ("gateway", "string", "gateway", "string"), ("client", "string", "client", "string"), ("target", "string", "target", "string"), ("request_processing_time", "string", "request_processing_time", "float"), ("target_processing_time", "string", "target_processing_time", "float"), ("response_processing_time", "string", "response_processing_time", "float"), ("elb_status_code", "string", "elb_status_code", "string"), ("target_status_code", "string", "target_status_code", "string"), ("received_bytes", "string", "received_bytes", "int"), ("sent_bytes", "string", "sent_bytes", "int"), ("target_request", "string", "target_request", "string"), ("user_agent", "string", "user_agent", "string"), ("ssl_cipher", "string", "ssl_cipher", "string"), ("ssl_protocol", "string", "ssl_protocol", "string"), ("target_group_arn", "string", "target_group_arn", "string"), ("trace_id", "string", "trace_id", "string"), ("domain_name", "string", "domain_name", "string"), ("chosen_cert_arn", "string", "chosen_cert_arn", "string"), ("matched_rule_priority", "string", "matched_rule_priority", "string"), ("timestamp_request", "string", "timestamp_request", "timestamp"), ("actions_executed", "string", "actions_executed", "string"), ("redirect_url", "string", "redirect_url", "string"), ("error_reason", "string", "error_reason", "string"), ("url_path", "string", "url_path", "string"), ("url_query", "string", "url_query", "string"), ("url_token", "string", "url_token", "string"), ("user_id", "string", "user_id", "string"), ("asset_group", "string", "asset_group", "string"), ("service_name", "string", "service_name", "string"), ("layer_names", "string", "layer_names", "string"), ("slip_layer_ids", "string", "slip_layer_ids", "string"), ("slip_organisation_id", "string", "slip_organisation_id", "string"), ("service_type", "string", "service_type", "string"), ("service_class", "string", "service_class", "string"), ("request_type", "string", "request_type", "string"), ("datadownload_ext", "string", "datadownload_ext", "string")], transformation_ctx = "applymapping1")
## @type: ResolveChoice
## @args: [choice = "make_struct", transformation_ctx = "resolvechoice2"]
## @return: resolvechoice2
## @inputs: [frame = applymapping1]
resolvechoice2 = ResolveChoice.apply(frame = applymapping1, choice = "make_struct", transformation_ctx = "resolvechoice2")
## @type: DropNullFields
## @args: [transformation_ctx = "dropnullfields3"]
## @return: dropnullfields3
## @inputs: [frame = resolvechoice2]
dropnullfields3 = DropNullFields.apply(frame = resolvechoice2, transformation_ctx = "dropnullfields3")
## @type: DataSink
## @args: [connection_type = "s3", connection_options = {"path": "s3://uat-logs-etl"}, format = "parquet", transformation_ctx = "datasink4"]
## @return: datasink4
## @inputs: [frame = dropnullfields3]
datasink4 = glueContext.write_dynamic_frame.from_options(frame = dropnullfields3, connection_type = "s3", connection_options = {"path": args['s3bucketETL']}, format = "parquet", transformation_ctx = "datasink4")
job.commit()
